File: 2026-02/src/day_04/boj_24445/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
vector<int> adj[100001];
bool visited[100001];
int order[100001];

void bfs(int start) {
    queue<int> q;
    q.push(start);

    visited[start] = true;

    int cnt = 1;

    while (!q.empty()) {
        int node = q.front();
        q.pop();

        order[node] = cnt++;

        for (int next : adj[node]) {
            if (visited[next]) continue;

            q.push(next);
            visited[next] = true;
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int m, r;
    cin >> n >> m >> r;

    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    for (int i = 1; i <= n; i++) {
        sort(adj[i].rbegin(), adj[i].rend());
    }

    bfs(r);

    for (int i = 1; i <= n; i++) {
        cout << order[i] << '\n';
    }
}

File: 2026-02/src/day_04/boj_24445/Main.java
package day_04.boj_24445;

import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static List<Integer>[] adj;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());

        adj = new ArrayList[1 + n];
        for (int i = 1; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            adj[u].add(v);
            adj[v].add(u);
        }

        for (int i = 1; i <= n; i++) {
            adj[i].sort(Comparator.reverseOrder());
        }

        System.out.println(bfs(r));
    }

    static String bfs(int start) {
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(start);

        boolean[] vis = new boolean[1 + n];
        vis[start] = true;

        int[] order = new int[1 + n];
        int cnt = 1;

        while (!q.isEmpty()) {
            int cur = q.poll();
            order[cur] = cnt++;

            for (int nxt : adj[cur]) {
                if (vis[nxt]) continue;

                q.offer(nxt);
                vis[nxt] = true;
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {
            sb.append(order[i]).append("\n");
        }

        return sb.toString();
    }
}
